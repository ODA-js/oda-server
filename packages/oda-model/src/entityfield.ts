import {
  RelationFieldBaseInput,
  RelationFieldBaseInternal,
  RelationFieldBaseMetaInfo,
  IRelationFieldBase,
  RelationFieldBase,
  RelationFieldBasePersistence,
  RelationFieldBaseOutput,
} from './relationfieldbase';
import { merge } from 'lodash';
import {
  EntityType,
  Nullable,
  assignValue,
  MetaModelType,
  Multiplicity,
} from './types';
import { HasMany } from './hasmany';
import { HasOne } from './hasone';
import { Internal, MetaData } from './element';

export interface IEntityField
  extends IRelationFieldBase<
    EntityFieldMeta,
    EntityFieldInput,
    EntityFieldPersistence,
    EntityFieldOutput
  > {
  readonly type: EntityType;
  readonly multiplicity: Multiplicity;
}

export interface EntityFieldPersistence extends RelationFieldBasePersistence {}

export interface EntityFieldMeta
  extends RelationFieldBaseMetaInfo<EntityFieldPersistence> {}

export interface EntityFieldInternal extends RelationFieldBaseInternal {
  type: EntityType;
  multiplicity: Multiplicity;
}

export interface EntityFieldInput
  extends RelationFieldBaseInput<EntityFieldMeta, EntityFieldPersistence> {
  type: EntityType;
  multiplicity?: Multiplicity;
}

export interface EntityFieldOutput
  extends RelationFieldBaseOutput<EntityFieldMeta, EntityFieldPersistence> {
  type: EntityType;
  multiplicity?: Multiplicity;
}

const defaultMetaInfo = {
  persistence: {
    derived: false,
    persistent: true,
    identity: false,
    indexed: false,
  },
};
const defaultInput = { metadata: defaultMetaInfo };

export class EntityField
  extends RelationFieldBase<
    EntityFieldMeta,
    EntityFieldInput,
    EntityFieldInternal,
    EntityFieldPersistence,
    EntityFieldOutput
  >
  implements IEntityField {
  public get modelType(): MetaModelType {
    return 'entity-field';
  }
  constructor(init: EntityFieldInput) {
    super(merge({}, defaultInput, init));
  }
  get type(): EntityType {
    return this[Internal].type;
  }

  get multiplicity(): Multiplicity {
    return this[Internal].multiplicity;
  }

  get derived(): boolean {
    return this[MetaData].persistence.derived;
  }

  public updateWith(input: Nullable<EntityFieldInput>) {
    super.updateWith(input);

    assignValue<
      EntityFieldInternal,
      EntityFieldInput,
      NonNullable<EntityFieldInput['multiplicity']>
    >({
      src: this[Internal],
      input,
      field: 'multiplicity',
      effect: (src, value) => {
        if (src.type) {
          src.type.multiplicity = value ? 'many' : 'one';
        } else {
          if (input.type) {
            input.type.multiplicity = value ? 'many' : 'one';
          }
        }
        src.multiplicity = value;
      },
      required: true,
      setDefault: src => {
        src.multiplicity = src.type ? src.type.multiplicity || 'one' : 'one';
      },
    });

    assignValue<
      EntityFieldInternal,
      EntityFieldInput,
      NonNullable<EntityFieldInput['type']>
    >({
      src: this[Internal],
      input,
      field: 'type',
      effect: (src, type) => {
        if (!type.multiplicity) {
          type.multiplicity = 'one';
        }
        switch (type.multiplicity) {
          case 'one': {
            src.relation = new HasOne({
              hasOne: `${type.name}#`,
              entity: this.metadata.entity,
              field: this.name,
              embedded: true,
            });
            src.multiplicity = 'one';
            break;
          }
          case 'many': {
            src.relation = new HasMany({
              hasMany: `${type.name}#`,
              entity: this.metadata.entity,
              field: this.name,
              embedded: true,
            });
            src.multiplicity = 'many';
            break;
          }
          default:
        }
        src.type = type;
      },
      required: true,
    });

    assignValue<
      EntityFieldMeta,
      EntityFieldInput,
      NonNullable<EntityFieldInput['identity']>
    >({
      src: this[MetaData],
      input,
      field: 'identity',
      allowEffect: () => false,
      required: true,
      setDefault: src => (src.persistence.identity = false),
    });

    assignValue<
      EntityFieldMeta,
      EntityFieldInput,
      NonNullable<EntityFieldInput['derived']>
    >({
      src: this[MetaData],
      input,
      field: 'derived',
      effect: (src, value) => (src.persistence.derived = value),
      required: true,
      setDefault: src => (src.persistence.derived = false),
    });

    assignValue<
      EntityFieldMeta,
      EntityFieldInput,
      NonNullable<EntityFieldInput['persistent']>
    >({
      src: this[MetaData],
      input,
      field: 'persistent',
      effect: (src, value) => (src.persistence.persistent = value),
      required: true,
      setDefault: src => (src.persistence.persistent = false),
    });
  }

  public toObject(): EntityFieldOutput {
    return merge({}, super.toObject(), {
      type: this[Internal].type,
      multiplicity: this[Internal].multiplicity,
    } as Partial<EntityFieldOutput>);
  }

  public mergeWith(payload: Nullable<EntityFieldInput>) {
    super.mergeWith(payload);
  }
}
