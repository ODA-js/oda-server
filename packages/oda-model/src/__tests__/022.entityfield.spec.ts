import 'jest';
import { EntityField } from '../entityfield';

describe('EntityField', () => {
  it('default', () => {
    const res = new EntityField({
      name: 'A',
      type: { type: 'entity', name: 'b' },
    });
    expect(res.type.multiplicity).toBe('one');
    expect(res.relation.verb).toBe('HasOne');
    expect(res.toObject()).toMatchSnapshot('default');
    expect(res).toMatchSnapshot('default');
  });
  it('default many', () => {
    const res = new EntityField({
      name: 'A',
      multiplicity: 'many',
      type: { type: 'entity', name: 'b' },
    });
    expect(res.type.multiplicity).toBe('many');
    expect(res.relation.verb).toBe('HasMany');
    expect(res.toObject()).toMatchSnapshot('default');
    expect(res).toMatchSnapshot('default');
  });
});
