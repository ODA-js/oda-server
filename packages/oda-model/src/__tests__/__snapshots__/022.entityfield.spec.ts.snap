// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityField default many: default 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
  "multiplicity": "many",
  "name": "a",
  "title": undefined,
  "type": Object {
    "multiplicity": "many",
    "name": "b",
    "type": "entity",
  },
}
`;

exports[`EntityField default many: default 2`] = `
EntityField {
  Symbol(internal): Object {
    "args": Map {},
    "multiplicity": "many",
    "name": "a",
    "relation": HasMany {
      Symbol(internal): Object {
        "field": "a",
        "hasMany": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "B",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "name": "HasManyAs",
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "HasManyAs",
          "normal": "As",
          "short": "As",
        },
        "persistence": Object {
          "embedded": true,
          "single": false,
          "stored": false,
        },
        "verb": "HasMany",
      },
    },
    "type": Object {
      "multiplicity": "many",
      "name": "b",
      "type": "entity",
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
}
`;

exports[`EntityField default: default 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
  "multiplicity": "one",
  "name": "a",
  "title": undefined,
  "type": Object {
    "multiplicity": "one",
    "name": "b",
    "type": "entity",
  },
}
`;

exports[`EntityField default: default 2`] = `
EntityField {
  Symbol(internal): Object {
    "args": Map {},
    "multiplicity": "one",
    "name": "a",
    "relation": HasOne {
      Symbol(internal): Object {
        "field": "a",
        "hasOne": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "B",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "name": "HasOneA",
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "HasOneA",
          "normal": "A",
          "short": "A",
        },
        "persistence": Object {
          "embedded": true,
          "single": true,
          "stored": false,
        },
        "verb": "HasOne",
      },
    },
    "type": Object {
      "multiplicity": "one",
      "name": "b",
      "type": "entity",
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
}
`;
